### ISSUE TITLE
vmware: disk controller "pvscsi" is not applied on scsiX.virtualDev

### ISSUE THREAD

**Issue by weizhouapache** [apache/cloudstack#5877]
in vmx file, the scsiX.virtualDev is not changed to "pvscsi" when root and data disk controller are set to "pvscsi"

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
vmware
~~~

##### CLOUDSTACK VERSION

~~~
4.16
~~~

##### SUMMARY

By default, the scsiX.virtualDev is "lsilogic".
when I set root and data disk controller to "pvscsi", I expect that the scsiX.virtualDev is changed to "pvscsi".

##### STEPS TO REPRODUCE

~~~
1. create a vm and start it

the vm has config in vmx file

# grep scsi /vmfs/volumes/85f59f3e-422b762e/i-2-4-VM/i-2-4-VM.vmx
scsi0.virtualDev = "lsilogic"
scsi0.present = "TRUE"
scsi1.virtualDev = "lsilogic"
scsi1.present = "TRUE"
scsi2.virtualDev = "lsilogic"
scsi2.present = "TRUE"
scsi3.virtualDev = "lsilogic"
scsi3.present = "TRUE"
scsi0.pciSlotNumber = "16"
scsi1.pciSlotNumber = "32"
scsi2.pciSlotNumber = "33"
scsi3.pciSlotNumber = "34"

2. stop the vm, and change  vm setting

rootDiskController=pvscsi
dataDiskController=pvscsi

3. start the vm


# grep scsi /vmfs/volumes/85f59f3e-422b762e/i-2-4-VM/i-2-4-VM.vmx
scsi0.virtualDev = "lsilogic"
scsi0.present = "TRUE"
scsi1.virtualDev = "lsilogic"
scsi1.present = "TRUE"
scsi2.virtualDev = "lsilogic"
scsi2.present = "TRUE"
scsi3.virtualDev = "lsilogic"
scsi3.present = "TRUE"
scsi0.pciSlotNumber = "16"
scsi1.pciSlotNumber = "32"
scsi2.pciSlotNumber = "33"
scsi3.pciSlotNumber = "34"
scsi0:0.deviceType = "scsi-hardDisk"
scsi0:0.fileName = "ROOT-4.vmdk"
scsi0:0.present = "TRUE"
dataDiskController = "pvscsi"
rootDiskController = "pvscsi"
scsi0:0.redo = ""

~~~

##### EXPECTED RESULTS

~~~

scsi0.virtualDev = "pvscsi"
scsi0.present = "TRUE"
scsi1.virtualDev = "pvscsi"
scsi1.present = "TRUE"
scsi2.virtualDev = "pvscsi"
scsi2.present = "TRUE"
scsi3.virtualDev = "pvscsi"
scsi3.present = "TRUE"

~~~

![image](https://user-images.githubusercontent.com/57355700/150085625-4621693d-d381-449c-91b2-f815927acaa6.png)

##### ACTUAL RESULTS

~~~
scsi0.virtualDev = "lsilogic"
scsi0.present = "TRUE"
scsi1.virtualDev = "lsilogic"
scsi1.present = "TRUE"
scsi2.virtualDev = "lsilogic"
scsi2.present = "TRUE"
scsi3.virtualDev = "lsilogic"
scsi3.present = "TRUE"

~~~
![image](https://user-images.githubusercontent.com/57355700/150085273-3f2290d7-320a-4d4d-9eff-bb63c12a15a5.png)

**Comment by rohityadavcloud** [comment]
@weizhouapache @sureshanaparti let's check VMwareResourceBase, it might be a historic limitation that under some circumstance controllers are not always added as configured in the template/disk controller properties.

**Comment by sureshanaparti** [comment]
> @weizhouapache @sureshanaparti let's check VMwareResourceBase, it might be a historic limitation that under some circumstance controllers are not always added as configured in the template/disk controller properties.

@rohityadavcloud Correct, pvscsi controller is bypassed to use lsilogic controller when the support for different scsi controller types was introduced. It seems the vmdk property "ddb.adapterType" is not properly set for pvscsi controller, and also it is not supported for virtual hardware version < 7. May be, we have to re-visit it and check whether the limitation still exists or not.

https://github.com/apache/cloudstack/blob/7f000f50164b2c47a726518f5752d7bdae68f057/vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VmdkAdapterType.java#L30-L31

https://github.com/apache/cloudstack/blob/7f000f50164b2c47a726518f5752d7bdae68f057/vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java#L708-L711

**Comment by sureshanaparti** [comment]
Hi @weizhouapache Have you noticed any logs indicating 'pvscsi' controller is bypassed to use 'lsilogic' controller?

**Comment by weizhouapache** [comment]
@sureshanaparti @rohityadavcloud 
no, I did not notice such logs.
I have investigated the issue and found that cloudstack does not update existing disk.

https://github.com/apache/cloudstack/blob/4.16.0.0/plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java#L1808-L1817

the workaround is
(1) create a vm, but not start it
(2) update vm settings, set root and data disk controller to 'pvscsi'
(3) start the vm
Then the vm will have correct SCSI controller.

**Comment by DaanHoogland** [comment]
fixed in #5910
